.github/workflows/ci.yml

```diff
@@ -0,0 +1,459 @@
+name: CI
+on:
+  workflow_call:
+    inputs:
+      # extra jobs must be multi-line string, as there's no support for `type: array` for inputs
+      extra_jobs:
+        type: string
+        required: false
+        default: ''
+      composer_require_extra:
+        type: string
+        required: false
+        default: ''
+      dynamic_matrix:
+        type: boolean
+        default: true
+      simple_matrix:
+        type: boolean
+        default: false
+      endtoend:
+        type: boolean
+        default: true
+      phpcoverage:
+        type: boolean
+        default: false
+      phplinting:
+        type: boolean
+        default: true
+      phpunit:
+        type: boolean
+        default: true
+      js:
+        type: boolean
+        default: true
+
+jobs:
+
+  context:
+    name: Context
+    runs-on: ubuntu-latest
+    env:
+      # Put this through an env variable to prevent possible string substitution injection via pull-request branch name
+      GITHUB_HEAD_REF: ${{ github.head_ref }}
+    steps:
+      - name: Context
+        run: |
+          # https://docs.github.com/en/actions/learn-github-actions/contexts#context-availability
+          # github.base_ref - The target branch of the pull request in a workflow run.
+          echo "github.base_ref: ${{ github.base_ref }}"
+          # github.head_ref - The source branch of the pull request in a workflow run.
+          echo "github.head_ref: $GITHUB_HEAD_REF"
+          # github.ref - The branch or tag ref that triggered the workflow run. 
+          # For branches this is the format refs/heads/<branch_name>, and for tags it is refs/tags/<tag_name>
+          echo "github.ref: ${{ github.ref }}"
+          # github.ref_name - The branch or tag name that triggered the workflow run - same as github.ref though without the leading refs/[heads|tags]/
+          echo "github.ref_name: ${{ github.ref_name }}"
+          # gitbub.repository - The owner and repository name. For example, Codertocat/Hello-World
+          echo "$github.repository: ${{ github.repository }}"
+
+  genmatrix:
+    name: Generate matrix
+    runs-on: ubuntu-latest
+
+    # gha-generate-matrix script.php will sanitise matrix outputs so they're safe to use within bash
+    outputs:
+      matrix: ${{ steps.generate-matrix.outputs.matrix }}
+    
+    steps:
+      - name: Generate matrix
+        id: generate-matrix
+        uses: emteknetnz/gha-generate-matrix@main
+        with:
+          extra_jobs: ${{ inputs.extra_jobs }}
+          simple_matrix: ${{ inputs.simple_matrix }}
+          endtoend: ${{ inputs.endtoend }}
+          phpcoverage: ${{ inputs.phpcoverage }}
+          phplinting: ${{ inputs.phplinting }}
+          phpunit: ${{ inputs.phpunit }}
+          js: ${{ inputs.js }}
+
+  tests:
+    needs: genmatrix
+
+    strategy:
+      # set fail-fast to false prevent one job from cancelling other jobs
```

Comment:
No, it's cancels at the job level

Reply:
So.... does this mean "If the `test` job fails, other jobs will still run" or does it mean "If other jobs fail, the `test` job will still run"?

or does this not refer to the jobs defined in this workflow, but the individual runs of this `test` job as defined in the matrix? If this is what's meant, please specify that.

Maybe it will be good to also include a link to the relevant docs so if people are still unsure they can very quickly just refer to that?

Reply:
Updated comment and linked to github docs